db/index.go

```diff
@@ -0,0 +1,495 @@
+// Copyright 2022 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	ds "github.com/ipfs/go-datastore"
+
+	"github.com/ipfs/go-datastore/query"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/errors"
+)
+
+const (
+	indexFieldValuePrefix = "v"
+	indexFieldNilValue    = "n"
+)
+
+type CollectionIndex interface {
+	Save(context.Context, datastore.Txn, *client.Document) error
+	Update(context.Context, datastore.Txn, *client.Document, *client.Document) error
+	RemoveAll(context.Context, datastore.Txn) error
+	Name() string
+	Description() client.IndexDescription
+}
+
+func getFieldValConverter(kind client.FieldKind) func(any) ([]byte, error) {
+	switch kind {
+	case client.FieldKind_STRING:
+		return func(val any) ([]byte, error) {
+			return []byte(val.(string)), nil
+		}
+	case client.FieldKind_INT:
+		return func(val any) ([]byte, error) {
+			intVal, ok := val.(int64)
+			if !ok {
+				return nil, errors.New("invalid int value")
+			}
+			return []byte(strconv.FormatInt(intVal, 10)), nil
+		}
+	case client.FieldKind_FLOAT:
+		return func(val any) ([]byte, error) {
+			floatVal, ok := val.(float64)
+			if !ok {
+				return nil, errors.New("invalid float value")
+			}
+			return []byte(strconv.FormatFloat(floatVal, 'f', -1, 64)), nil
+		}
+	case client.FieldKind_BOOL:
+		return func(val any) ([]byte, error) {
+			boolVal, ok := val.(bool)
+			if !ok {
+				return nil, errors.New("invalid bool value")
+			}
+			var intVal int64 = 0
+			if boolVal {
+				intVal = 1
+			}
+			return []byte(strconv.FormatInt(intVal, 10)), nil
+		}
+	case client.FieldKind_DATETIME:
+		return func(val any) ([]byte, error) {
+			timeStrVal, ok := val.(string)
+			if !ok {
+				return nil, errors.New("invalid datetime value")
+			}
+			_, err := time.Parse(time.RFC3339, timeStrVal)
+			if err != nil {
+				return nil, err
+			}
+			return []byte(timeStrVal), nil
+		}
+	default:
+		panic("there is no test for this case")
+	}
+}
+
+func NewCollectionIndex(
+	collection client.Collection,
+	desc client.IndexDescription,
+) CollectionIndex {
+	index := &collectionSimpleIndex{collection: collection, desc: desc}
+	schema := collection.Description().Schema
+	fieldID := schema.GetFieldKey(desc.Fields[0].Name)
+	field := schema.Fields[fieldID]
```

Comment:
todo: FieldID is not necessarily the slice index. Please do not rely on this. There is a function on collection description `GetFieldByID` that you should use.  John might be moving this onto schema in his fetcher PR.